[Security Solution][Detections] adds bulk edit rule actions

vitaliidm opened the pull request:
## Summary

- addresses https://github.com/elastic/security-team/issues/2072
- adds new bulk edit actions: `add_rule_actions`, `set_rule_actions`
- moved immutability check from rule `validateMutatedParams` to action validator. Because, rule immutability depends on actions performed on it, not only on `immutable` property
- adds some test coverage
- using workaround for https://github.com/elastic/kibana/issues/139084, by muting/unmuting single rule. This would only happen:
    - if rule was muted before, throttle set to some value
    - rule was unmuted, throttle set to `no_actions`

### Feature recording
Note: callouts on recording are not up to date

https://user-images.githubusercontent.com/92328789/185381912-6c4a25f6-fb36-4c31-bf08-8ec28f2358c0.mov

### Screen 
<img width="1465" alt="Screenshot 2022-08-25 at 17 23 56" src="https://user-images.githubusercontent.com/92328789/186731607-574687b8-8a7a-43de-8f30-6cda3dcecfc5.png">

### Checklist

Delete any items that are not applicable to this PR.

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

### Release note

Adding bulk edit of rule actions

vitaliidm commented on x-pack/plugins/security_solution/public/detections/containers/detection_engine/rules/validate_rule_actions_field/validate_rule_actions_field.ts:
```diff
@@ -59,34 +51,3 @@ export const validateRuleActionsField =
     }
   };
-
-export const getSchema = ({
```
moved to `x-pack/plugins/security_solution/public/detections/components/rules/step_rule_actions/get_schema.ts`

elasticmachine commented:
Pinging @elastic/security-detections-response (Team:Detections and Resp)

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

tylersmalley commented on packages/kbn-optimizer/limits.yml:
```diff
@@ -93,7 +93,7 @@ pageLoadAssetSize:
   interactiveSetup: 80000
   expressionTagcloud: 27505
-  securitySolution: 273763
+  securitySolution: 301655
```
What is being done to address this? The ask to reduce this number was done [here](https://github.com/elastic/kibana/issues/95870). This is loaded on **every** page refresh, affecting the overall experience of Kibana. We have [documentation](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html#ci-metric-resolving-overages) to assist with reducing these bundle sizes through code splitting.

vitaliidm replied:
@tylersmalley, I wasn't aware of the task you mentioned. After quick look, also found a tech debt task where bundle size should be investigated: https://github.com/elastic/security-team/issues/2441 

Do you consider current limit raise as critical and blocker for this PR and prefer to be handled in it, instead of handling bundle size in dedicated tickets?

vitaliidm replied:
I was able to resolve it and revert limit increase.
Thanks for bringing attention to this

tylersmalley replied:
Much appreciated, thanks! ❤️

marshallmain reviewed (approved):
Codeowner review only, alerts area changes LGTM

jpdjere commented on x-pack/plugins/security_solution/server/lib/detection_engine/rules/bulk_edit_rules.ts:
```diff
@@ -41,12 +50,73 @@ export const bulkEditRules = ({
+
+  if (rulesAction) {
+    const errors: BulkEditError[] = [];
+    const rules = await pMap(
```
Just trying to understand the use of `pMap` here. My understanding from the codebase is that we need to respect the `MAX_RULES_TO_UPDATE_IN_PARALLEL` to prevent performance issues and timeouts.

However, I've seen the elsewhere in the code that the `initPromisePool` method from `x-pack/plugins/security_solution/server/utils/promise_pool.ts` is used instead of `pMap`. Any reason not use use that instead?

vitaliidm replied:
If I remember correctly `initPromisePool` was introduced to make efficient calls in parallel, i.e. what is pMap doing.
pMap is used widely across Kibana and even in Security Solution, thus my choice of using it.
initPromisePool also have some additional logic of wrapping executed callbacks and returning them in a certain format, which is not needed for my case.

I guess, eventually we would like to replace initPromisePool, if there is no drawback of using pMap where can be used.

jpdjere commented:
Other than a question posted above, code changes good look to me.

Nevertheless, I'd like to go through the test cases for this feature before approving.
@banderror just saw in the thread with Karan Verma that the test cases [have been added to this document.](https://docs.google.com/document/d/1QiNkakZqCtxYirwR5CGGIXfFOKhvMdraFah-K2KOCs0/edit#)
Are those cases OK to follow and fully test the feature?

vitaliidm commented:
> @banderror just saw in the thread with Karan Verma that the test cases [have been added to this document.](https://docs.google.com/document/d/1QiNkakZqCtxYirwR5CGGIXfFOKhvMdraFah-K2KOCs0/edit#)
Are those cases OK to follow and fully test the feature?

Those are drafts from QAs that would need yet to be reviewed.
After review and finalizing, we will add missing scenarios in follow-up PR.

xcrzx commented:
I tested locally and noticed that applying bulk actions to prebuilt rules leads to their version increment. The version should stay intact after bulk operations. Otherwise, it would break the rules upgrade process.

vitaliidm commented:
>I tested locally and noticed that applying bulk actions to prebuilt rules leads to their version increment. The version should stay intact after bulk operations. Otherwise, it would break the rules upgrade process.

@xcrzx, thanks for catching this. Fixed and added some tests to cover it 👍

banderror commented:
Found a bug while testing locally: some of the "rule variables" are duplicated in the list.

<img width="1206" alt="Screenshot 2022-08-30 at 19 50 58" src="https://user-images.githubusercontent.com/7359339/187508595-1bbae789-7061-4a8e-8166-2e506821a682.png">

banderror commented:
As a result of bulk adding rule actions, we show the toast that explains how bulk editing index patterns handles rules with data views. This is confusing since bulk editing rule actions has nothing to do with index patterns or data views.

<img width="382" alt="Screenshot 2022-08-30 at 20 06 36" src="https://user-images.githubusercontent.com/7359339/187511500-c3d36988-7aee-4f78-8d37-992f5b96fb92.png">

banderror commented:
If I create a rule with `Perform no actions` and bulk add _0 actions_ but set frequency to let's say `Weekly`, this rule will not be updated. At the same time, rules with some frequency and actions will be.

Steps to repro:

1. Import rules using the ndjson below. One of them - the `No actions` rule - has `throttle: "no_actions"` and `actions: []`. Other rules have `throtthe: "rule"` or `throtthe: "1h"`.
2. Bulk add rule actions. In the form, select `Weekly` but don't add any actions.
3. Click Save.
4. Observe that the `No actions` rule's `throttle` hasn't changed and is still equal to `no_actions`. Other rules have been updated and how have `throttle: "7d"`.

Ndjson for repro:

```
{"id":"72d362c0-2889-11ed-b135-5faec038f000","updated_at":"2022-08-30T17:36:48.072Z","updated_by":"elastic","created_at":"2022-08-30T17:30:32.053Z","created_by":"elastic","name":"Hourly - 1 action","tags":[],"interval":"5m","enabled":false,"description":"-","risk_score":21,"severity":"low","license":"","output_index":"","meta":{"from":"1m","kibana_siem_app_url":"http://localhost:5601/kbn/app/security"},"author":[],"false_positives":[],"from":"now-360s","rule_id":"c6dba132-4159-4f83-8931-e3f3b631f235","max_signals":100,"risk_score_mapping":[],"severity_mapping":[],"threat":[],"to":"now","references":[],"version":3,"exceptions_list":[],"immutable":false,"related_integrations":[],"required_fields":[],"setup":"","type":"query","language":"kuery","index":["apm-*-transaction*","auditbeat-*","endgame-*","filebeat-*","logs-*","packetbeat-*","traces-apm*","winlogbeat-*","-*elastic-cloud-logs-*"],"query":"host.name:*","filters":[],"throttle":"1h","actions":[{"group":"default","id":"1095b270-2889-11ed-b135-5faec038f000","params":{"body":"{{context.rule.name}}\nAction 1"},"action_type_id":".webhook"}]}
{"id":"eaaf1740-2888-11ed-b135-5faec038f000","updated_at":"2022-08-30T17:36:48.076Z","updated_by":"elastic","created_at":"2022-08-30T17:26:43.639Z","created_by":"elastic","name":"On each rule execution - 1 action","tags":[],"interval":"5m","enabled":false,"description":"-","risk_score":21,"severity":"low","license":"","output_index":"","meta":{"from":"1m","kibana_siem_app_url":"http://localhost:5601/kbn/app/security"},"author":[],"false_positives":[],"from":"now-360s","rule_id":"d1033dc3-9950-4dea-b93a-4ecb8707b546","max_signals":100,"risk_score_mapping":[],"severity_mapping":[],"threat":[],"to":"now","references":[],"version":2,"exceptions_list":[],"immutable":false,"related_integrations":[],"required_fields":[],"setup":"","type":"query","language":"kuery","index":["apm-*-transaction*","auditbeat-*","endgame-*","filebeat-*","logs-*","packetbeat-*","traces-apm*","winlogbeat-*","-*elastic-cloud-logs-*"],"query":"host.name:*","filters":[],"throttle":"rule","actions":[{"group":"default","id":"1095b270-2889-11ed-b135-5faec038f000","params":{"body":"{{context.rule.name}}\nAction 1"},"action_type_id":".webhook"}]}
{"id":"e4804bf0-2888-11ed-b135-5faec038f000","updated_at":"2022-08-30T17:36:48.071Z","updated_by":"elastic","created_at":"2022-08-30T17:26:33.269Z","created_by":"elastic","name":"No actions","tags":[],"interval":"5m","enabled":false,"description":"-","risk_score":21,"severity":"low","license":"","output_index":"","meta":{"from":"1m","kibana_siem_app_url":"http://localhost:5601/kbn/app/security"},"author":[],"false_positives":[],"from":"now-360s","rule_id":"c1a4a016-3c71-4385-922c-d842b0c9c822","max_signals":100,"risk_score_mapping":[],"severity_mapping":[],"threat":[],"to":"now","references":[],"version":1,"exceptions_list":[],"immutable":false,"related_integrations":[],"required_fields":[],"setup":"","type":"query","language":"kuery","index":["apm-*-transaction*","auditbeat-*","endgame-*","filebeat-*","logs-*","packetbeat-*","traces-apm*","winlogbeat-*","-*elastic-cloud-logs-*"],"query":"host.name:*","filters":[],"throttle":"no_actions","actions":[]}
{"id":"8ff852c0-2889-11ed-b135-5faec038f000","updated_at":"2022-08-30T17:36:48.069Z","updated_by":"elastic","created_at":"2022-08-30T17:31:20.951Z","created_by":"elastic","name":"Hourly - 2 actions","tags":[],"interval":"5m","enabled":false,"description":"-","risk_score":21,"severity":"low","license":"","output_index":"","meta":{"from":"1m","kibana_siem_app_url":"http://localhost:5601/kbn/app/security"},"author":[],"false_positives":[],"from":"now-360s","rule_id":"7a414a22-4cd5-4fdd-a0c4-cef508942c0d","max_signals":100,"risk_score_mapping":[],"severity_mapping":[],"threat":[],"to":"now","references":[],"version":4,"exceptions_list":[],"immutable":false,"related_integrations":[],"required_fields":[],"setup":"","type":"query","language":"kuery","index":["apm-*-transaction*","auditbeat-*","endgame-*","filebeat-*","logs-*","packetbeat-*","traces-apm*","winlogbeat-*","-*elastic-cloud-logs-*"],"query":"host.name:*","filters":[],"throttle":"1h","actions":[{"group":"default","id":"1095b270-2889-11ed-b135-5faec038f000","params":{"body":"{{context.rule.name}}\nAction 1"},"action_type_id":".webhook"},{"group":"default","id":"1095b270-2889-11ed-b135-5faec038f000","params":{"body":"{{context.rule.name}}\nAction 2"},"action_type_id":".webhook"}]}
{"id":"2b3cd130-2889-11ed-b135-5faec038f000","updated_at":"2022-08-30T17:36:48.070Z","updated_by":"elastic","created_at":"2022-08-30T17:28:31.949Z","created_by":"elastic","name":"On each rule execution - 2 actions","tags":[],"interval":"5m","enabled":false,"description":"-","risk_score":21,"severity":"low","license":"","output_index":"","meta":{"from":"1m","kibana_siem_app_url":"http://localhost:5601/kbn/app/security"},"author":[],"false_positives":[],"from":"now-360s","rule_id":"192cb271-e0d2-4738-b3e6-33fcdf57e8b3","max_signals":100,"risk_score_mapping":[],"severity_mapping":[],"threat":[],"to":"now","references":[],"version":3,"exceptions_list":[],"immutable":false,"related_integrations":[],"required_fields":[],"setup":"","type":"query","language":"kuery","index":["apm-*-transaction*","auditbeat-*","endgame-*","filebeat-*","logs-*","packetbeat-*","traces-apm*","winlogbeat-*","-*elastic-cloud-logs-*"],"query":"host.name:*","filters":[],"throttle":"rule","actions":[{"group":"default","id":"1095b270-2889-11ed-b135-5faec038f000","params":{"body":"{{context.rule.name}}\nAction 1"},"action_type_id":".webhook"},{"group":"default","id":"1095b270-2889-11ed-b135-5faec038f000","params":{"body":"{{context.rule.name}}\nAction 2"},"action_type_id":".webhook"}]}
{"exported_count":5,"exported_rules_count":5,"missing_rules":[],"missing_rules_count":0,"exported_exception_list_count":0,"exported_exception_list_item_count":0,"missing_exception_list_item_count":0,"missing_exception_list_items":[],"missing_exception_lists":[],"missing_exception_lists_count":0}
```

banderror commented:
Hey @vitaliidm, I will continue testing and reviewing it tomorrow. I posted a few bugs that I found so far.

banderror commented:
Hey @elastic/security-docs folks, if you have some time, could you please check the new bulk editing form for the wording of callouts and labels? This screenshot could be enough but if you'd like to click through it we can deploy the PR to Cloud.

<img width="1205" alt="Screenshot 2022-08-30 at 20 41 11" src="https://user-images.githubusercontent.com/7359339/187517629-76f841e4-9105-479f-b7ed-7d4f829e15bc.png">

nastasha-solomon commented:
@banderror thanks for pinging us about this! It'd be great to play around with this in a cloud deployment if you don't mind spinning one up.

vitaliidm commented:
@banderror 

> Found a bug while testing locally: some of the "rule variables" are duplicated in the list.

Thanks for catching this: already pushed a fix

> As a result of bulk adding rule actions, we show a toast that explains how bulk editing index patterns handles rules with data views. This is confusing since bulk editing rule actions has nothing to do with index patterns or data views.

This is a behaviour that present in main, and this message is getting displayed for all actions: tags, index, timeline.
Was introduced in 8.4.0 in https://github.com/elastic/kibana/pull/130929
We may try to limit showing this message only when users apply index actions to rules? I can create a ticket for this if you on board with this idea.

>  If I create a rule with Perform no actions and bulk add 0 actions but set frequency to let's say Weekly, this rule will not be updated. At the same time, rules with some frequency and actions will be.

This is an expected behaviour, which I even covered in tests.
If rule has no actions, it will always have throttle as 'no_actions'. And it makes sense: there is no action to perform, so no frequency is needed to be set.
Here is a logic on our side that handles it:
https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rules/utils.ts#L105-L107

``typescript
   if (rule.muteAll || rule.actions.length === 0) {
      return NOTIFICATION_THROTTLE_NO_ACTIONS;
    } else if (
```

where NOTIFICATION_THROTTLE_NO_ACTIONS === `no_actions`

I've created a simple graph while working on PR to track logic around rule actions properties: https://docs.google.com/drawings/d/1qim8mMkdZjwwboNM0XEQ0QqAKye7Qq3IJ4M1iJTb3wI/edit

banderror commented:
> This is a behaviour that present in main, and this message is getting displayed for all actions: tags, index, timeline.
Was introduced in 8.4.0 in https://github.com/elastic/kibana/pull/130929
We may try to limit showing this message only when users apply index actions to rules? I can create a ticket for this if you on board with this idea.

Yes, I think we should open a ticket for that. Please do @vitaliidm, thanks!

> This is an expected behaviour, which I even covered in tests.
If rule has no actions, it will always have throttle as 'no_actions'. And it makes sense: there is no action to perform, so no frequency is needed to be set.

Ahhh I see, right, that makes sense. Could be a little confusing for the users who expect that action frequency will be applied to rules in any case. Maybe we should explain it in the callout.

vitaliidm commented:
@elasticmachine merge upstream

banderror commented:
@nastasha-solomon Alrighty, so the cloud deployment is ready. Please let me know if you have trouble logging into it. 
Credentials: https://p.elstc.co/paste/rR6D4Sgu#FzC3nfo0TNveTP4K2YGwmZMgzYQzhUr7xl48kYzz1vG

banderror reviewed (commented):
Reviewed the changes and it LGTM 👍 
Left a few comments, most of them are minor and nits.

Tested locally and wasn't able to find any other issues.
Thank you for such a big effort @vitaliidm!

vitaliidm commented on x-pack/plugins/security_solution/common/detection_engine/schemas/request/perform_bulk_action_schema.test.ts:
```diff
@@ -373,5 +373,163 @@ describe('perform_bulk_action_schema', () => {
+        expect(message.schema).toEqual({});
+      });
+
+      test('invalid request: invalid action_type_id property in actions array', () => {
+        const payload = {
+          query: 'name: test',
+          action: BulkAction.edit,
+          [BulkAction.edit]: [
+            {
+              type: BulkActionEditType.add_rule_actions,
+              value: {
+                throttle: '1h',
+                actions: [
+                  {
+                    action_type_id: '.webhook',
+                    group: 'default',
+                    id: '458a50e0-1a28-11ed-9098-47fd8e1f3345',
```
`actionTypeId` is omitted for rule actions payload on rulesClient side for update and for bulkEdit

```typescript
type NormalizedAlertAction = Omit<RuleAction, 'actionTypeId'>;


export interface UpdateOptions<Params extends RuleTypeParams> {
  id: string;
  data: {
    name: string;
    tags: string[];
    schedule: IntervalSchedule;
    actions: NormalizedAlertAction[];
    params: Params;
    throttle: string | null;
    notifyWhen: RuleNotifyWhenType | null;
  };
}
```
That's why bulk edit schema defined without it, this field is not used in rulesClient.
It seems, passing it in a single update might be redundant.

Added a comment regarding this format in request schema

banderror replied:
> It seems, passing it in a single update might be redundant.

Got it. Looks like `action_type_id` is neither needed for creating a rule:

```ts
export interface CreateOptions<Params extends RuleTypeParams> {
  data: Omit<
    Rule<Params>,
    | 'id'
    | 'createdBy'
    | 'updatedBy'
    | 'createdAt'
    | 'updatedAt'
    | 'apiKey'
    | 'apiKeyOwner'
    | 'muteAll'
    | 'mutedInstanceIds'
    | 'actions'
    | 'executionStatus'
    | 'snoozeSchedule'
    | 'isSnoozedUntil'
  > & { actions: NormalizedAlertAction[] };
  options?: {
    id?: string;
    migrationVersion?: Record<string, string>;
  };
}

export interface UpdateOptions<Params extends RuleTypeParams> {
  id: string;
  data: {
    name: string;
    tags: string[];
    schedule: IntervalSchedule;
    actions: NormalizedAlertAction[];
    params: Params;
    throttle: string | null;
    notifyWhen: RuleNotifyWhenType | null;
  };
}
```

> Added a comment regarding this format in request schema

Thanks 👍

vitaliidm commented on x-pack/plugins/security_solution/public/detections/pages/detection_engine/rules/all/bulk_actions/forms/rule_actions_form.tsx:
```diff
@@ -0,0 +1,216 @@
+        color="primary"
+        data-test-subj="bulkEditRulesRuleActionInfo"
+        title={
+          <FormattedMessage
+            id="xpack.securitySolution.detectionEngine.rules.allRules.bulkActions.edit.addRuleActions.infoCalloutTitle"
+            defaultMessage="Add actions for {rulesCount, plural, one {# rule} other {# rules}} you’ve selected. Here some tips and suggestions:"
+            values={{ rulesCount }}
+          />
+        }
+      >
+        <ul>
+          <li>
+            <FormattedMessage
+              id="xpack.securitySolution.detectionEngine.rules.allRules.bulkActions.edit.addRuleActions.actionFrequencyDetail"
+              defaultMessage="Action frequency is shared to all new and existing actions."
+            />
+          </li>
+          <li>
+            <FormattedMessage
+              id="xpack.securitySolution.detectionEngine.rules.allRules.bulkActions.edit.addRuleActions.deleteActionsDetail"
+              defaultMessage='If you’d like to delete actions for all selected rules, select "Perform no actions" in the dropdown and check Overwrite all selected rule actions.'
+            />
+          </li>
```
Using `FormattedMessage` is something that expected in most of the cases https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/GUIDELINE.md#usage-of-appropriate-component

Had `RULE_VARIABLES_DETAIL` in translations as FormattedMessage component for some reason didn't work with encoded curly braces symbols. So I had to resort to using of `i18n.translate`

banderror replied:
@vitaliidm FWIW I'd expect `i18n.translate()` to be used most of the time, because usually what we need is a simple localized string. When we need something more, like a custom localized markup, then `FormattedMessage` might help as it supports some customizations like a `children` function etc.

The guide says:

> In the case where the string is expected (aria-label, placeholder), Call JS function i18n.translate() from the@kbn/i18n package.

In other words, by default we extract translations into a separate file and use this syntax:

```ts
  THROTTLE_HELP_TEXT: i18n.translate(
    'xpack.securitySolution.detectionEngine.rules.allRules.bulkActions.edit.addRuleActions.throttleHelpText',
    {
      defaultMessage:
        'Select when automated actions should be performed if a rule evaluates as true.',
    }
  ),
```

banderror replied:
> I think we would need to align on approach here within a team

Sounds good, I added this to our next Tech agenda 👍

banderror replied:
Alright, so everyone thinks that `i18n.translate` should be used by default.

vitaliidm commented on x-pack/plugins/security_solution/public/detections/pages/detection_engine/rules/helpers.tsx:
```diff
@@ -361,14 +361,44 @@ export const redirectToDetections = (
 
-const getRuleSpecificRuleParamKeys = (ruleType: Type) => {
-  const queryRuleParams = ['index', 'filters', 'language', 'query', 'saved_id'];
+const commonRuleParamsKeys = [
+  'id',
+  'name',
+  'description',
+  'false_positives',
+  'rule_id',
+  'max_signals',
+  'risk_score',
+  'output_index',
+  'references',
+  'severity',
+  'timeline_id',
+  'timeline_title',
+  'threat',
+  'type',
+  'version',
+];
+const queryRuleParams = ['index', 'filters', 'language', 'query', 'saved_id'];
+const machineLearningRuleParams = ['anomaly_threshold', 'machine_learning_job_id'];
+const thresholdRuleParams = ['threshold', ...queryRuleParams];
+
+const getAllRuleParamsKeys = (): string[] => {
+  const allRuleParamsKeys = [
+    ...commonRuleParamsKeys,
+    ...queryRuleParams,
+    ...machineLearningRuleParams,
+    ...thresholdRuleParams,
+  ].sort();
 
+  return Array.from(new Set<string>(allRuleParamsKeys));
+};
```
created a ticket for it https://github.com/elastic/kibana/issues/140019

banderror replied:
Thank you

vitaliidm commented on x-pack/plugins/security_solution/server/lib/detection_engine/rules/bulk_edit_rules.ts:
```diff
@@ -41,12 +50,73 @@ export const bulkEditRules = ({
+    .reverse()
+    .find(({ type }) =>
+      [BulkActionEditType.set_rule_actions, BulkActionEditType.add_rule_actions].includes(type)
+    ) as BulkActionEditPayloadRuleActions;
```
refactored a bit:
- removed type casting
- remove reverse in favour of accessing a last item of filtered results

vitaliidm commented on x-pack/test/detection_engine_api_integration/security_and_spaces/group1/perform_bulk_action.ts:
```diff
@@ -43,6 +47,31 @@ export default ({ getService }: FtrProviderContext): void => {
+  const fetchRuleByAlertApi = (ruleId: string) =>
+    supertest.get(`/api/alerting/rule/${ruleId}`).set('kbn-xsrf', 'true');
+
+  const createWebHookAction = async () =>
+    (
+      await supertest
+        .post('/api/actions/action')
+        .set('kbn-xsrf', 'true')
+        .send(getWebHookAction())
+        .expect(200)
```
Let's leave it for now as it is, as it seems a bit out of PR scope( `getWebHookAction` is used in multiple test files of different methods)

It uses `/api/actions/action` endpoint.
Then in rules actions object,  it stores id of that webHookAction as `id` property
 ```
              {
                ...webHookActionMock,
                id: hookAction.id,
                action_type_id: '.webhook',
              },
```

Might be even more confusing: created connector id equals to action id

vitaliidm commented on x-pack/plugins/security_solution/common/detection_engine/schemas/common/schemas.ts:
```diff
@@ -379,55 +379,6 @@ export enum BulkActionEditType {
   'set_index_patterns' = 'set_index_patterns',
   'set_timeline' = 'set_timeline',
+  'add_rule_actions' = 'add_rule_actions',
+  'set_rule_actions' = 'set_rule_actions',
```
Moved them to `x-pack/plugins/security_solution/common/detection_engine/schemas/request/perform_bulk_action_schema.ts`

vitaliidm commented:
@banderror 
Have addressed the comments

Also created a couple of tickets:
- https://github.com/elastic/kibana/issues/139897
- https://github.com/elastic/kibana/issues/140019
- https://github.com/elastic/kibana/issues/140031

banderror commented on x-pack/plugins/security_solution/server/lib/detection_engine/rules/bulk_edit_rules.ts:
```diff
@@ -41,12 +50,73 @@ export const bulkEditRules = ({
+        try {
+          if (rule.muteAll && rulesAction.value.throttle !== NOTIFICATION_THROTTLE_NO_ACTIONS) {
+            await rulesClient.unmuteAll({ id: rule.id });
+            return (await readRules({ rulesClient, id: rule.id, ruleId: undefined })) ?? rule;
```
@vitaliidm this function has a `return null;` statement in 3 other cases as well (so overall 4 cases). If I'm not mistaken all of them can be reduced to the following:

- rule with the given `id` or `ruleId` doesn't exist / not found
- rule exists but its `alertTypeId` is not one of the Security Solution's rule types

If `bulkEdit` succeeded for a rule, and then `readRules` returns null for it, I think it means that the rule was deleted between the 2 calls, right? Are you handling this race condition with `??`? If so, it LGTM.

banderror reviewed (approved):
@vitaliidm Thank you for addressing the comments and creating the follow-up tickets. I have 2 comments left but none of them seem to be blocking.

Before you merge this PR please assist @elastic/security-docs with anything they might need for reviewing the wording used in the new bulk editing form.

joepeeples reviewed (commented):
Cool new feature, thanks for adding docs team for review! Added a few line edits. 

Also, overall we might want to reconsider the name "Add rule actions" (in the "Bulk actions" menu, the flyout title, etc.), because a user might also use this UI to edit or remove actions, not just add. Maybe just the name "Rule actions" would cover all those use cases.

vitaliidm commented on x-pack/plugins/security_solution/public/detections/pages/detection_engine/rules/all/bulk_actions/translations.tsx:
```diff
@@ -55,3 +55,42 @@ export const bulkApplyTimelineTemplate = {
+    'xpack.securitySolution.detectionEngine.rules.allRules.bulkActions.edit.addRuleActions.ruleVariablesDetail',
+    {
+      defaultMessage:
+        'Rule variables may only affect some of the rules that you selected based on the rule types(for example \\u007b\\u007bcontext.rule.threshold\\u007d\\u007d will only work for threshold rule type).',
```
@joepeeples 
> For example, for non-threshold type rules, will {{context.rule.threshold}} appear as literal text? or empty space? or error/fail?

It will appear as empty space

joepeeples replied:
Thanks @vitaliidm — how about "will only display values..." instead of "will only work..."?

joepeeples commented on x-pack/plugins/security_solution/public/detections/pages/detection_engine/rules/all/bulk_actions/translations.tsx:
```diff
@@ -55,3 +55,42 @@ export const bulkApplyTimelineTemplate = {
+    'xpack.securitySolution.detectionEngine.rules.allRules.bulkActions.edit.addRuleActions.ruleVariablesDetail',
+    {
+      defaultMessage:
+        'Rule variables may only affect some of the rules that you selected based on the rule types(for example \\u007b\\u007bcontext.rule.threshold\\u007d\\u007d will only work for threshold rule type).',
```
```suggestion
        'Rule variables may only affect some of the rules that you selected based on the rule types(for example \\u007b\\u007bcontext.rule.threshold\\u007d\\u007d will only display values for threshold rule type).',
```

joepeeples replied:
```suggestion
        'Rule variables may affect only some of the rules you select, based on the rule types (for example, \\u007b\\u007bcontext.rule.threshold\\u007d\\u007d will only display values for threshold rules).',
```

vitaliidm commented on x-pack/plugins/security_solution/public/detections/pages/detection_engine/rules/all/bulk_actions/forms/rule_actions_form.tsx:
```diff
@@ -0,0 +1,216 @@
+          <li>
+            <FormattedMessage
+              id="xpack.securitySolution.detectionEngine.rules.allRules.bulkActions.edit.addRuleActions.deleteActionsDetail"
+              defaultMessage='If you’d like to delete actions for all selected rules, select "Perform no actions" in the dropdown and check Overwrite all selected rule actions.'
```
Actually, we can. I've changed it to bold version

vitaliidm commented:
Thanks for your help, @joepeeples 

Here is the final version:
- without list
- marked 
<img width="1355" alt="Screenshot 2022-09-06 at 17 03 58" src="https://user-images.githubusercontent.com/92328789/188683505-0ec5f33f-5e5f-48c7-b9fb-d523e39e9085.png">
menu and button as *bold*

> Also, overall we might want to reconsider the name "Add rule actions" (in the "Bulk actions" menu, the flyout title, etc.), because a user might also use this UI to edit or remove actions, not just add. Maybe just the name "Rule actions" would cover all those use cases.

This name is in line with other actions menu:
Tags -> Add tags -> User can overwrite tags by clicking Overwrite checkbox
Index patterns -> Add Index patterns -> User can overwrite patterns by clicking Overwrite checkbox

joepeeples commented:
> Here is the final version:
> 
> * without list
> * marked menu and button as _bold_

@vitaliidm Thanks for the updated screenshot! Actually I think it's OK to use a bulleted list for the tips, and they probably work better that way; I was just saying that it wasn't necessary to introduce the list the way the original heading was doing. 

> This name is in line with other bulk edit actions menu: Tags -> Add tags -> User can overwrite tags by clicking Overwrite checkbox Index patterns -> Add Index patterns -> User can overwrite patterns by clicking Overwrite checkbox

That makes sense to me, thanks for the extra context.

vitaliidm commented:
@joepeeples 
> I think it's OK to use a bulleted list for the tips, and they probably work better that way

👍 , got the list back
<img width="1333" alt="Screenshot 2022-09-06 at 18 12 54" src="https://user-images.githubusercontent.com/92328789/188698545-b13b0303-c8d1-48bd-ae4f-86d213d5eab1.png">

joepeeples reviewed (approved):
LGTM!
